main.py

```diff
@@ -0,0 +1,347 @@
+    return images.get_serving_url(blob_key)
+
+
+# Returns the url of the original photo.
+def get_original(photo_name, generation):
+    original_photo = '{}{}{}{}{}{}'.format(
+        'https://storage.googleapis.com/',
+        PHOTO_BUCKET,
+        '/',
+        photo_name,
+        '?generation=',
+        generation)
+    return original_photo
+
+
+# Shrinks specified photo to thumbnail size and returns resulting thumbnail.
+def create_thumbnail(photo_name):
+    filename = '{}{}{}{}'.format(
+        '/gs/',
+        PHOTO_BUCKET,
+        '/',
+        photo_name)
+    image = images.Image(filename=filename)
+    image.resize(width=180, height=200)
+    return image.execute_transforms(output_encoding=images.JPEG)
+
+
+# Stores thumbnail in GCS bucket under name thumbnail_key.
+def store_thumbnail_in_gcs(thumbnail_key, thumbnail):
+    write_retry_params = cloudstorage.RetryParams(
+        backoff_factor=1.1,
+        max_retry_period=15)
+    filename = '{}{}{}{}'.format(
+        '/',
+        THUMBNAIL_BUCKET,
+        '/',
+        thumbnail_key)
+    with cloudstorage.open(filename, 'w',
+                           retry_params=write_retry_params) as filehandle:
+        filehandle.write(thumbnail)
+
+
+# Deletes thumbnail from GCS bucket and deletes thumbnail_reference from
+# datastore.
+def delete_thumbnail(thumbnail_key):
+    filename = '{}{}{}{}'.format(
+        '/gs/',
+        THUMBNAIL_BUCKET,
+        '/',
+        thumbnail_key)
+    blob_key = blobstore.create_gs_key(filename)
+    images.delete_serving_url(blob_key)
+    thumbnail_reference = ThumbnailReference.query(
+        ThumbnailReference.thumbnail_key == thumbnail_key).get()
+    thumbnail_reference.key.delete()
+
+    filename = '{}{}{}{}'.format(
+        '/',
+        THUMBNAIL_BUCKET,
+        '/',
+        thumbnail_key)
+    cloudstorage.delete(filename)
+
+
+# Use Cloud Vision API to get labels for a photo.
+def get_labels(uri, photo_name):
+    service = googleapiclient.discovery.build('vision', 'v1')
+    labels = []
+
+    # Label photo with its name, sans extension.
+    index = photo_name.index(".jpg")
+    photo_name_label = photo_name[:index]
+    labels.append(photo_name_label)
+
+    service_request = service.images().annotate(body={
+        'requests': [{
+            'image': {
+                'source': {
+                    'imageUri': uri
+                }
+            },
+            'features': [{
+                'type': 'LABEL_DETECTION',
+                'maxResults': MAX_LABELS
+            }]
+        }]
+    })
+    response = service_request.execute()
+    labels_full = response['responses'][0].get('labelAnnotations')
+
+    ignore = ['of', 'like', 'the', 'and', 'a', 'an', 'with']
+
+    # Add labels to the labels list if they are not already in the list and are
+    # not in the ignore list.
+    if labels_full is not None:
+        for label in labels_full:
+            if label['description'] not in labels:
```

Comment:
You could just use a `set`.

Reply:
Hi! Thanks for taking the time to review our stuff. `ignore` is now a set, but `labels` is stored as a repeated string property in Datastore and so cannot be a set.

Reply:
you can use a set for the purpose of calculation and then cast it to a list for storage.

Reply:
Sounds good. Done.